File: forecasting/presentation.py
"""Presentation schema — the agent ↔ TUI contract for Market Models.

A Presentation is a versioned header plus an ordered list of typed render
"blocks". The forecasting agent emits one (via the ``emit_market_presentation``
tool); the TUI renders each block by ``type`` with a tolerant fallback so an
unknown/extended type never crashes the view. ``validate_presentation`` runs
before persistence to keep the contract honest.

This module is deliberately dependency-light (stdlib only) so it can be imported
by the gateway, the tools, and tests without pulling in numpy/agent machinery.
"""

from __future__ import annotations

from typing import Any

SCHEMA_VERSION = "market-presentation-v1"

# Block types the renderer knows how to draw. Unknown types still render (title +
# note + raw JSON) but validation flags them so we notice drift early.
BLOCK_TYPES: frozenset[str] = frozenset(
    {
        "narrative",
        "finding",
        "metric",
        "timeseries",
        "scatter",
        "regression",
        "bars",
        "table",
        "fan",
        "scenario",
        "assumptions",
        "sources",
    }
)

_STATUSES = frozenset({"complete", "partial", "failed"})

# Per-type required fields (beyond the common {type,id}). Kept intentionally
# small — the renderer is defensive, so we only require what it cannot invent.
_REQUIRED_FIELDS: dict[str, tuple[str, ...]] = {
    "narrative": ("body",),
    "finding": ("claim",),
    "metric": ("label", "value"),
    "timeseries": ("series",),
    "scatter": ("points",),
    "regression": ("r2", "coeffs"),
    "bars": ("series",),
    "table": ("columns", "rows"),
    "fan": ("x", "median"),
    "scenario": ("scenarios",),
    "assumptions": ("items",),
    "sources": ("items",),
}


def new_block_id(prefix: str, index: int) -> str:
    """Stable, human-readable block id (no randomness — replayable)."""
    return f"{prefix}-{index}"


def validate_presentation(obj: Any) -> tuple[bool, list[str]]:
    """Return ``(ok, errors)``. Non-raising: callers decide how to handle.

    Enforces the structural contract: a dict header with a ``blocks`` list whose
    entries each have a known ``type`` + their required fields. ``ok`` is True
    only when there are zero errors; ``errors`` always lists every problem so a
    self-repair loop can fix them in one pass.
    """
    errors: list[str] = []

    if not isinstance(obj, dict):
        return False, ["presentation must be a JSON object"]

    if not str(obj.get("title") or "").strip():
        errors.append("missing title")

    status = obj.get("status")
    if status is not None and status not in _STATUSES:
        errors.append(f"status must be one of {sorted(_STATUSES)}; got {status!r}")

    blocks = obj.get("blocks")
    if not isinstance(blocks, list):
        return False, errors + ["blocks must be a list"]

    seen_ids: set[str] = set()
    for i, block in enumerate(blocks):
        where = f"blocks[{i}]"
        if not isinstance(block, dict):
            errors.append(f"{where} must be an object")
            continue

        btype = block.get("type")
        if not isinstance(btype, str) or not btype:
            errors.append(f"{where} missing type")
            continue
        if btype not in BLOCK_TYPES:
            errors.append(f"{where} unknown type {btype!r} (renders as fallback)")

        bid = block.get("id")
        if isinstance(bid, str) and bid:
            if bid in seen_ids:
                errors.append(f"{where} duplicate id {bid!r}")
            seen_ids.add(bid)

        for field in _REQUIRED_FIELDS.get(btype, ()):  # type: ignore[arg-type]
            if field not in block or block.get(field) in (None, ""):
                errors.append(f"{where} ({btype}) missing required field {field!r}")

    return (len([e for e in errors if "renders as fallback" not in e]) == 0), errors


def build_presentation(
    *,
    model_id: str,
    version: int,
    title: str,
    question: str,
    blocks: list[dict],
    summary: str = "",
    status: str = "complete",
    as_of_analysis: str | None = None,
    as_of_data: str | None = None,
    data_refs: list[str] | None = None,
    model_run_refs: list[str] | None = None,
    diagnostics: dict | None = None,
) -> dict[str, Any]:
    """Assemble a normalized Presentation header. Ensures every block has an id."""
    normalized: list[dict] = []
    for i, block in enumerate(blocks or []):
        if not isinstance(block, dict):
            continue
        b = dict(block)
        if not b.get("id"):
            b["id"] = new_block_id(str(b.get("type") or "block"), i)
        normalized.append(b)

    return {
        "schema_version": SCHEMA_VERSION,
        "model_id": model_id,
        "version": version,
        "title": title,
        "question": question,
        "summary": summary,
        "status": status if status in _STATUSES else "complete",
        "as_of_analysis": as_of_analysis,
        "as_of_data": as_of_data,
        "blocks": normalized,
        "data_refs": list(data_refs or []),
        "model_run_refs": list(model_run_refs or []),
        "diagnostics": dict(diagnostics or {}),
    }


def sanitize_presentation_prose(presentation: dict) -> dict:
    """Apply house-style text cleanup to prose-bearing blocks in place-ish.

    Reuses ``forecasting.writeup.sanitize_writeup_text`` (the canonical em-dash /
    whitespace enforcer) on narrative bodies, finding claims, and notes. Returns
    the same dict for chaining. Best-effort: if writeup import fails, returns the
    presentation untouched.
    """
    try:
        from forecasting.writeup import sanitize_writeup_text
    except Exception:
        return presentation

    for block in presentation.get("blocks", []):
        if not isinstance(block, dict):
            continue
        for key in ("body", "claim", "note", "title", "caption"):
            if isinstance(block.get(key), str):
                block[key] = sanitize_writeup_text(block[key])
    if isinstance(presentation.get("summary"), str):
        presentation["summary"] = sanitize_writeup_text(presentation["summary"])
    return presentation

File: ui-tui/src/lib/presentation.ts
// The Market Models Presentation contract (mirrors forecasting/presentation.py).
// The gateway returns untyped JSON; `normalizePresentation` defensively coerces
// it into this shape (parity with asRpcResult's posture). The renderer dispatches
// per block `type` with a tolerant fallback, so unknown/extended types never
// crash the view.

export interface XYPoint {
  label?: string
  x: number
  y: number
}

export interface Band {
  lower: number[]
  p_hi: number
  p_lo: number
  upper: number[]
}

// Optional typed helpers for the @hermes/viz-rendered block types (DX only — the
// renderer reads from the loose PresentationBlock and never requires these).
export interface HeatmapBlock {
  colLabels?: string[]
  diverging?: boolean
  matrix: number[][]
  max?: number
  min?: number
  rowLabels?: string[]
  type: 'heatmap'
}

export interface FanBlock {
  bands?: { lower: number[]; p_hi?: number; p_lo?: number; upper: number[] }[]
  median: number[]
  paths?: number[][]
  type: 'fan'
  x?: number[]
}

// Blocks are intentionally loose: `type` is a string and fields are optional, so
// the renderer degrades gracefully on partial/extended payloads.
export interface PresentationBlock {
  [key: string]: unknown
  id?: string
  note?: string
  title?: string
  type: string
}

export interface Presentation {
  as_of_analysis?: null | string
  as_of_data?: null | string
  blocks: PresentationBlock[]
  diagnostics?: Record<string, unknown>
  model_id?: string
  question?: string
  schema_version?: string
  status?: string
  summary?: string
  title: string
  version?: number
}

export interface MarketModelListItem {
  current_version?: number
  depth?: string
  id: string
  // The latest presentation's build status (complete | partial | failed).
  last_status?: string
  question?: string
  status?: string
  tags?: string[]
  title: string
  updated_at?: string
}

const asString = (v: unknown, fallback = ''): string => (typeof v === 'string' ? v : fallback)

/** Coerce a raw gateway value into a Presentation, or null if unusable. */
export const normalizePresentation = (raw: unknown): null | Presentation => {
  if (!raw || typeof raw !== 'object') {
    return null
  }

  const obj = raw as Record<string, unknown>
  const rawBlocks = Array.isArray(obj.blocks) ? obj.blocks : []

  const blocks: PresentationBlock[] = rawBlocks
    .filter((b): b is Record<string, unknown> => Boolean(b) && typeof b === 'object' && typeof (b as Record<string, unknown>).type === 'string')
    .map((b, i) => ({ ...(b as PresentationBlock), id: asString((b as PresentationBlock).id, `block-${i}`) }))

  return {
    as_of_analysis: typeof obj.as_of_analysis === 'string' ? obj.as_of_analysis : null,
    as_of_data: typeof obj.as_of_data === 'string' ? obj.as_of_data : null,
    blocks,
    diagnostics: obj.diagnostics && typeof obj.diagnostics === 'object' ? (obj.diagnostics as Record<string, unknown>) : {},
    model_id: asString(obj.model_id),
    question: asString(obj.question),
    schema_version: asString(obj.schema_version),
    status: asString(obj.status, 'complete'),
    summary: asString(obj.summary),
    title: asString(obj.title, 'Untitled model'),
    version: typeof obj.version === 'number' ? obj.version : undefined
  }
}

/** Coerce a raw list payload into model list items. */
export const normalizeModelList = (raw: unknown): MarketModelListItem[] => {
  const rows = raw && typeof raw === 'object' && Array.isArray((raw as Record<string, unknown>).models)
    ? ((raw as Record<string, unknown>).models as unknown[])
    : Array.isArray(raw)
      ? (raw as unknown[])
      : []

  return rows
    .filter((m): m is Record<string, unknown> => Boolean(m) && typeof m === 'object' && typeof (m as Record<string, unknown>).id === 'string')
    .map(m => ({
      current_version: typeof m.current_version === 'number' ? m.current_version : 0,
      depth: asString(m.depth),
      id: asString(m.id),
      last_status: asString(m.last_status),
      question: asString(m.question),
      status: asString(m.status, 'active'),
      tags: Array.isArray(m.tags) ? (m.tags as string[]).filter(t => typeof t === 'string') : [],
      title: asString(m.title, 'Untitled model'),
      updated_at: asString(m.updated_at)
    }))
}
